### ISSUE TITLE
Packaging

### ISSUE THREAD

**Issue by pricelessbrewing** [beerjson/beerjson#51]
1) Any references for additions should be packaging instead of bottling. It's more general. Miscellaneous ingredient additions reference bottling, submitting a pull for this today, but see below for a necessary discussion. 

2) Should packaging be listed as a fermentation step, if not does it need to be included as a separate data structure? 

3) OTOH things that are often added to packaging, miscellaneous ingredients, hops, cultures, and can be used to indicate packaging form (350mL bottles, kegs, 750mL etc). Packaging date, (on that note, "created" date in recipe should explicitly state what it means (is it brew day, or packaging day?)

Also there should be an optional recipe parameter for serving temperature.

**Comment by rmsrosa** [comment]
I agree with 1. Point 2 is a good question, since we can prime the bottle, or even the keg. In the keg, we can even do more hop and other ingredient additions. I need to think more about that. Same for 3.

**Comment by pricelessbrewing** [comment]
>brew date vs created 

The current beerxml refers to brew date, but I think it would be a good idea to change "created" to "brew_date", and add an additional property of "packaged_date".

**Comment by krutilin** [comment]
Agree "brew_date" and "packaged_date" are useful in real life. And I use these two dates on the beer label on the bottle.

some offtopic:
Generally, my brewing workflow is following:
1) Decide what I want to brew.
2) Prepare and calculate recipe. Usually, it is beersmith tool.
3) Rewrite all steps in my paper notebook.
4) Do brew session, log steps, times, params to the paper notebook. And log photos to the Instagram. (but I want to use beerjson and maybe even git for this task, paper is not ideal storage)
5) Do fermenting and bottling sessions according to dates in my paper notebook and inst.

**Comment by rmsrosa** [comment]
I studied the json schema structure and I finally have a reasonable understanding of the beerjson schema itself. Now I think I can comment on point 2. And I noticed a few related problems. 

Ingredients such as hop, misc, and culture, for instance, can be added at many different stages that the current json does not allow for explicit identification. 

For instance, hop "use" accepts "dry hop" but it does not specify in which fermentation step. 

Misc accepts "fermentation" and "packaging", but it does not indicate the fermentation step, and packaging is not a specific step yet.

Cultures can be added at any fermentation stage, but if we want to prime at bottling adding another culture, this is not covered. 

Instead of adding packaging as a separate object in the recipe, I propose to replace "fermentation" with "cold_side", and "fermentation_step" by "cold_side_step". Then each cold side step should get an integer to identify the step and a title that the user can name however he wants ("fermentation", "conditioning", "centrifugue", "kegging", "bottling", "packaging", "barrel aging", whatever). 

Then, in each ingredient, the cold_condition_step of the addition can be referenced by the identifying number. Each software, of course, may allow the user to specify the step by its name, and the software translates to identifying number when saving the recipe as a beerjson file. We can also add a nonrequired field to allow the software to record the name of the step, as well, for better direct reading of the json recipe file, although the beerjson schema validation will not check whether the names and ids match.

What do you think. I can implement the changes, if you agree, now that I understand the schema.

**Comment by pricelessbrewing** [comment]
>Ingredients such as hop, misc, and culture, for instance, can be added at many different stages that the current json does not allow for explicit identification.

>Cultures can be added at any fermentation stage, but if we want to prime at bottling adding another culture, this is not covered.

Relevant issue: #44 add comments there. 

>Instead of adding packaging as a separate object in the recipe, I propose to replace "fermentation" with "cold_side", and "fermentation_step" by "cold_side_step". Then each cold side step should get an integer to identify the step and a title that the user can name however he wants ("fermentation", "conditioning", "centrifugue", "kegging", "bottling", "packaging", "barrel aging", whatever).

Just semantics, but this is kinda support atm via a new fermentation step and using the "vessel" property. IE the last step would be step X: name: kegged. Vessel: Corny keg. 

I'd like a distinct step schema for packaging, as it would clarify some things, and we could add things like carbonation, aging time, serving temperature, I'm not sure how I feel about "cold_side" though. Placing a packagingsteptype inside the fermentation.json file would be sufficient for me as it would be understood. Similarly, an "aging" step is understood to be part of the fermentation process even if actual sugar-> alcohol fermentation is not occurring.

**Comment by rmsrosa** [comment]
Sure, thanks for pointing me to issue #44, I should add the comment there.

As for "cold_side", I do like it very much, as well as the freedom that it gives to the brewer of naming the steps inside it whatever they want, but if everyone else likes the idea of shoving non-fermentative steps into "fermentation", that is fine with me, no big deal. Let us move on.

**Comment by pricelessbrewing** [comment]
Cold side would work, if mash is also changed to hot side. Would be self consistent, and it makes sense since not all "fermentation steps" are actually undergoing fermentation. IE aging, barrel aging, packaging, "secondary". 

Similarly in hot side it would be good IMO to explicitly state boiling, and whirl pooling as a concrete step object rather than being implied through the hop and kettle additions, for the purpose of instruction generating if nothing else.

**Comment by rmsrosa** [comment]
Since no one else gave any opinion about this, I think that, at this point, we could stick with fermentation_step encompassing every step before packaging and having packaging as a separate step. If we later decide to change that, it won't affect much the structure of the packaging step itself. So I suggest that @pricelessbrewing could start working on the schema for packaging since he already seems to have a good idea about it.

**Comment by pricelessbrewing** [comment]
@beerjson/beerjson-wg  I want to start on this, but still waiting on some input before proceeding. Should this be an object in the (A) overall recipe file (precedent being created date, and carbonation), (B) part of fermentation.json (precedent being aging steps are located in fermentation) file, or should it be a separate file package.json(C)?

**Comment by krutilin** [comment]
Vote for C, (But not "package.json" but "packaging.json", because package.json is name for project configuration file in nodejs.)

**Comment by pricelessbrewing** [comment]
As mentioned in another issue, and @rmsrosa mentioned above that I overlooked initially. 

@beerjson/beerjson-wg should procedure steps get an explicit index number? This would increase readibility and the recipe.json file can then include an explicit ordered process object containing all steps.

**Comment by pricelessbrewing** [comment]
Could someone review https://github.com/pricelessbrewing/beerjson/pull/3 please? Generic and real tests need to be made. 

I can't get around a travis samples.test.js fail. I think it might be because it's testing against https://github.com/beerjson/beerjson/ rather than my fork, but no idea how these work really.

**Comment by pricelessbrewing** [comment]
Still having issues with the travis CI tests. If someone could take a look that would be appreciated.

**Comment by pricelessbrewing** [comment]
@rmsrosa @krutilin I'd like to work on these new data structures, package.json and boil/whirlpool.json process steps but I can't get around the travis CI test mentioned above while updating snapshots or running the npm tests since it's a new file type I think.

> FAIL  tests\samples.test.js
>   ● Test suite failed to run
> 
>     can't resolve reference packaging.json#/definitions/PackagingProcedureType from id https://raw.githubusercontent.com/beerjson/beerjson/master/json/beer.json#

Could you please review my branch above, at which point I'll continue.

**Comment by krutilin** [comment]
@pricelessbrewing  will check "Test suite failed to run" on your branch today. On my side.

**Comment by pricelessbrewing** [comment]
I haven't kept it up to date since I haven't gotten the tests to pass the error above. 

I'll see if I can update that branch.
